### ISSUE TITLE
Helmfile is using yaml v2 encoders which is causing helmfile write-values output to be non deterministic

### ISSUE THREAD

**Issue by bgiegel** [helmfile/helmfile#1548]
### Operating system

Mac OS, Linux Centos stream9

### Helmfile Version

v0.165.0

### Helm Version

v3.15.1

### Bug description

I discovered a weird issue while trying to deploy an helmfile config to prod I noticed that even without any changes of my config helm diff was reporting a lot of changes. This is annoying because it means that release will be constantly re-deployed on each helmfile apply.

I then noticed the differences came from the fact that the order of the maps in my  yaml config was changing all the time for no reason and only for with specific alphanumerical strings like "4fe2ed584743bf9e" for example. I also noticed that the issue  appeared already at the write-values step.

I suspected a bug on the helmfile side. I therefore opened a discussion on helmfile github repository here : https://github.com/helmfile/helmfile/discussions/1547

@yxxhero kindly answered and asked me to provide more details. I created a small repo to reproduce the bug : https://github.com/bgiegel/helmfile-map-sorting-problem

I also started debugging the helmfile write-values command and found out that the issue was really happening when the merged values yaml find was encoded and written to disk. So the problem seemed to be linked to the yaml encoder. I digged a bit more and found out that helmfile was using yaml v2 golang lib, and that the yaml v3 api was available. So I decided to modify the code quickly to use the new yaml encoder v3 and tested it with my test repository. Turns out it fixes the issue. The maps order in the yaml output stay the same accross multiple helmfile commands.

You can find more info on yaml v3 here : https://ubuntu.com/blog/api-v3-of-the-yaml-package-for-go-is-available
I’m not sure exactly which addition fixed the issue but it might be this one : 
```
This has now changed. When v3 finds a map where all the keys are strings, which is the most common scenario by far, it will automatically decode it into a map[string]interface{} type`
```
yaml v3 is basically a rewrite of the API with nice additions like the possibily to encode decode yaml comments, manipulation of anchors and aliases, etc.

Note : By default the default indentation of yaml v3 uses 4 spaces. But there is a possibility to set a custom indentation apparently. Tell me if you want to continue to use the previous 2 spaces indent.

I already started to work on a fix, I’ll submit a PR when it’s ready.

### Example helmfile.yaml

I provided a full example that you can use to reproduce the error : https://github.com/bgiegel/helmfile-map-sorting-problem

### Error message you've seen (if any)

No errors

### Steps to reproduce

https://github.com/bgiegel/helmfile-map-sorting-problem

### Working Helmfile Version

This issue happens on the latest version and all the sub version. Not a regression.

### Relevant discussion

https://github.com/helmfile/helmfile/discussions/1547

**Comment by bgiegel** [comment]
Some updates. I’ve made some more tests. I made a commit on a local branch with helmfile switched completely to the new v3. I’ve tested it on my inital prod helmfile config to see if it solves the issue. And sadly it is not... 

The `write-values` command is fixed. It always sorts the maps entries in the same order. So that’s cool. 

The `template` command is not fixed it is still sorting the map entries differently on each execution... Which means it’s not enough. So I need to do more debugging.

Also, I noticed yaml v3 encoder is now checking for duplicate. Which is really nice since it allowed me to remove some of the duplicates in my prod config. But also it’s a breaking change so config that worked previously will not work after the fix. I don’t know how you want to handle this.

Lastly I’ll probably set the default spacing to 2 spaces because it seems weird. But also please tell me what you think @yxxhero (or other maintainers)

**Comment by bgiegel** [comment]
So... I did more testing. And it looks like you were right @yxxhero helm also seems to have the same issue when we pass multiple value files to the command, which is typically the use case of helm file. I checked the values file that helmfile give to the helm command and they are well sorted. But then helm probably merge the 2 values and do another yaml encoding with the same version of the API.... So it’s more complicated that initially anticipated (like always 😅).

Maybe they already know the issue and have a workaround... It would be surprising that I’m the first one to notice this on a project like helm... Or maybe they are stuck with v2 because of backward compatibility...
